plantcv/plantcv/transform/warp.py

```diff
@@ -1,58 +1,82 @@
 # warp image
 
 import cv2
+import skimage
 import os
 import numpy as np
-from plantcv.plantcv import print_image
-from plantcv.plantcv import plot_image
+import copy
+from plantcv.plantcv import params
+from plantcv.plantcv._debug import _debug
 from plantcv.plantcv import fatal_error
 from plantcv.plantcv import color_palette
-from plantcv.plantcv import params
 from plantcv.plantcv.visualize import overlay_two_imgs
+from plantcv.plantcv.transform import rescale
+from skimage import img_as_ubyte
 
+def _preprocess_img_dtype(img):
+    """ Transform the input image such that the datatype after transformation is uint8, ready for opencv functions
+    :param img: numpy.ndarray
+    :return: img_: numpy.ndarray
+    """
+    debug_mode = params.debug
+    params.debug = None
+    try:
+        img_         = rescale(img)
+    except RuntimeError:
+        img_ = img_as_ubyte(img)
+    params.debug = debug_mode
+    return img_
 
 def warp(img, refimg, pts, refpts, method='default'):
-    """Warp an image to another perspective
+    """Calculate a projective transform from 4 pairs of the corresponding points between image and reference image,
+    then performs the projective transformation on input image to align the input image to reference image
 
     Inputs:
-    img = grayscale or binary image data to be warped
-    refimg = RGB or grayscale image data to be used as reference
-    pts = 4 coordinates on img1
-    refpts = 4 coordinates on img2
-    method = method of finding the transformation. 'default', 'ransac', 'lmeds', 'rho'
+    img = image to be warped
+    refimg = image to be used as reference
+    pts = 4 coordinates on img
+    refpts = 4 coordinates on refimg
+    method = robust estimation algorithm when calculating projective transformation. 'default', 'ransac', 'lmeds', 'rho'
     Returns:
     warped_img = warped image
+    mat = transformation matrix
 
     :param img: numpy.ndarray
     :param refimg: numpy.ndarray
     :param pts: list of tuples
     :param refpts: list of tuples
     :param method: str
+    :return mat: numpy.ndarray
     :return warped_img: numpy.ndarray
+    :return mat: numpy.ndarray
     """
 
-    params.device += 1
+    if len(pts) != len(refpts):
+        fatal_error('Please provide same number of corresponding coordinates.')
 
-    if len(pts) != 4 or len(refpts) != 4:
```

Comment:
this function will fail if the number of points given is less than 4 so we should check for that